default/AI/CombatRatingsAI.py

```diff
@@ -0,0 +1,365 @@
+        :type hashable: bool
+        :return: list of ship stats
+        :rtype: list
+        """
+        return map(lambda x: x.get_stats(hashable=hashable), self.__ship_stats)
+
+    def get_ship_combat_stats(self):
+        """Returns list of ShipCombatStats of fleet."""
+        return list(self.__ship_stats)
+
+    def get_rating(self, enemy_stats=None):
+        """Calculates the rating of the fleet by combining all its ships ratings.
+
+        :param enemy_stats: enemy to be rated against
+        :type enemy_stats: ShipCombatStats
+        :return: Rating of the fleet
+        :rtype: float
+        """
+        return combine_ratings_list(map(lambda x: x.get_rating(enemy_stats), self.__ship_stats))
+
+    def __get_stats_from_fleet(self):
+        """Calculate fleet combat stats (i.e. the stats of all its ships)."""
+        universe = fo.getUniverse()
+        fleet = universe.getFleet(self.__fleet_id)
+        if not fleet:
+            return
+        for ship_id in fleet.shipIDs:
+            self.__ship_stats.append(ShipCombatStats(ship_id, self._consider_refuel))
+
+
+def get_fleet_rating(fleet_id, enemy_stats=None):
+    """Get rating for the fleet against specified enemy.
+
+    :param fleet_id: fleet to be rated
+    :type fleet_id: int
+    :param enemy_stats: enemy to be rated against
+    :type enemy_stats: ShipCombatStats
+    :return: Rating
+    :rtype: float
+    """
+    return FleetCombatStats(fleet_id, consider_refuel=False).get_rating(enemy_stats)
+
+
+def get_piloting_grades(species_name):
+    """Get species modifier.
+
+    :param species_name:
+    :type species_name: str
+    :return: 3 strings: weapons_grade, shield_grade, attacktroops_grade
+    """
+    if species_name not in piloting_grades:
+        spec_tags = []
+        if species_name:
+            species = fo.getSpecies(species_name)
+            if species:
+                spec_tags = species.tags
+            else:
+                print "Error: get_piloting_grades couldn't retrieve species '%s'" % species_name
+        piloting_grades[species_name] = (get_ai_tag_grade(spec_tags, 'WEAPONS'),
+                                         get_ai_tag_grade(spec_tags, 'SHIELDS'),
+                                         get_ai_tag_grade(spec_tags, 'ATTACKTROOPS'),
+                                         )
+    return piloting_grades[species_name]
+
+
+def weight_attack_troops(troops, grade):
+    """Re-weights troops on a ship based on species piloting grade.
+
+    :type troops: float
+    :type grade: str
+    :return: piloting grade weighted troops
+    :rtype: float
+    """
+    weight = {'NO': 0.0, 'BAD': 0.5, '': 1.0, 'GOOD': 1.5, 'GREAT': 2.0, 'ULTIMATE': 3.0}.get(grade, 1.0)
+    return troops * weight
+
+
+def weight_shields(shields, grade):
+    """Re-weights shields based on species defense bonus."""
+    offset = {'NO': 0, 'BAD': 0, '': 0, 'GOOD': 1.0, 'GREAT': 0, 'ULTIMATE': 0}.get(grade, 0)
```

Comment:
Please add comment that we have only single shield modification for species and does not know other values.  This dict values looks like an issue.